File: TopAnalysis/test/topAnalysis.py
#! /usr/bin/env python
import os
import glob
import math

from optparse import OptionParser

parser = OptionParser()


############################################
#            Job steering                  #
############################################

# Input inputFiles to use. This is in "glob" format, so you can use wildcards.
# If you get a "cannot find file" type of error, be sure to use "\*" instead
# of "*" to make sure you don't confuse the shell. 
parser.add_option('--inputFiles', metavar='F', type='string', action='store',
                  default = "",
                  dest='inputFiles',
                  help='Input files')

parser.add_option('--txtfiles', metavar='F', type='string', action='store',
                  default = "",
                  dest='txtfiles',
                  help='Input txt files')

parser.add_option("--onEOSCMS", action='store_true',
                  default=False,
                  dest='onEOSCMS',
                  help='onEOSCMS')

# Output name to use. 
parser.add_option('--outputFile', metavar='F', type='string', action='store',
                  default='top_fwlite.root',
                  dest='outputFile',
                  help='output file name')

# Using MC info or not. For MC, truth information is accessed.
parser.add_option('--doMC', metavar='F', action='store_true',
                  default=False,
                  dest='doMC',
                  help='Check MC Information')

# Which lepton type to use
parser.add_option('--lepType', metavar='F', type='int', action='store',
                  default=0,
                  dest='lepType',
                  help='Lepton type. Options are 0 = muons, 1 = electrons')

# Invert MET cut?
parser.add_option('--invertMET', action='store_true',
                  default=False,
                  dest='invertMET',
                  help='Invert MET cut')

# Invert PF isolation cut?
parser.add_option('--invertPFIso', action='store_true',
                  default=False,
                  dest='invertPFIso',
                  help='Invert PF isolation cut')


(options, args) = parser.parse_args()

argv = []

# Import everything from ROOT
import ROOT
ROOT.gROOT.Macro("rootlogon.C")

# Import stuff from FWLite
import sys
from DataFormats.FWLite import Events, Handle

#infile = open( options.inputFiles )
#infileStr = infile.read().rstrip()

#print 'Getting files from this dir: ' + infileStr

# Get the file list. 
#files = glob.glob( infileStr )

# Get the file list.
if options.inputFiles:
    files = glob.glob( options.files )
    print 'getting files', files
elif options.txtfiles:
    files = []
    with open(options.txtfiles, 'r') as input_:
        for line in input_:
            files.append(line.strip())
else:
    files = []

print 'getting files: ', files

if options.onEOSCMS:
        files = ["root://eoscms.cern.ch/eos/cms" + x for x in files]
        #print 'new files', *files, sep='\n'
        #print('new files', files[0], files[1], ..., sep='\n')

fname = options.txtfiles
fileN = fname[fname.rfind('/')+1:]

# Create the output file. 
f = ROOT.TFile(options.outputFile, "recreate")
f.cd()

# Make histograms
print "Creating histograms"
secvtxMassHist = ROOT.TH1F('secvtxMassHist', "Secondary Vertex Mass", 150, 0., 5.0)
secvtxMassHistB = ROOT.TH1F('secvtxMassHistB', "Secondary Vertex Mass, b jets", 150, 0., 5.0)
secvtxMassHistC = ROOT.TH1F('secvtxMassHistC', "Secondary Vertex Mass, c jets", 150, 0., 5.0)
secvtxMassHistL = ROOT.TH1F('secvtxMassHistL', "Secondary Vertex Mass, udsg jets", 150, 0., 5.0)
metVsIso = ROOT.TH2F('metVsIso', 'MET Versus PFIsolation', 15, 0., 150., 125, 0., 2.5)
jetPtHist = ROOT.TH1F('jetPtHist', 'Jet p_{T}', 150, 0., 600.)
m3Hist = ROOT.TH1F('m3Hist', 'M3 Histogram', 150, 0., 600.)

############################################
# Physics level parameters for systematics #
############################################

# Kinematic cuts:
jetPtMin = 30.0
leadJetPtMin = 30.0
isoMax = 0.2
ssvheCut = 1.74
minJets = 4

if options.lepType == 0 :
    muonPtMin = 45.0
    electronPtMin = 20.0
    metMin = 20.0
    lepStr = 'Mu'
else:
    muonPtMin = 20.0
    electronPtMin = 35.0
    metMin = 20.0
    lepStr = 'Ele'


events = Events (files)

# Make the entirety of the handles required for the
# analysis. 

puHandle         = Handle( "std::vector<double>" )
puLabel    = ( "pileupWeight",   "", "CAT" )
jetHandle         = Handle( "std::vector<cat::Jet>" )
jetLabel = ("catJets","","CAT")
muonHandle         = Handle( "std::vector<cat::Muon>" )
muonLabel = ("catMuons","","CAT")
electronHandle         = Handle( "std::vector<cat::Electron>" )
electronLabel = ("catElectrons","","CAT")
metHandle = Handle( "std::vector<cat::MET>" )
metLabel = ("catMETs","","CAT")

# Keep some timing information
nEventsAnalyzed = 0
nEventsPassed4Jets = 0
nEventsPassed1Tag = 0
timer = ROOT.TStopwatch()
timer.Start()

pairs = []

# loop over events
count = 0
ntotal = events.size()
percentDone = 0.0
ipercentDone = 0
ipercentDoneLast = -1
print "Start looping"
for event in events:
    nEventsAnalyzed += 1
    ipercentDone = int(percentDone)
    if ipercentDone != ipercentDoneLast :
        ipercentDoneLast = ipercentDone
        print 'Processing {0:10.0f}/{1:10.0f} : {2:5.0f}%'.format(
            count, ntotal, ipercentDone )
    count = count + 1
    percentDone = float(count) / float(ntotal) * 100.0


    ################################################
    #   Retrieve the jet four-vector
    #   ------------------------------------
    #      The jet 4-vectors are large and hence
    #      take a long time to read out. If you don't
    #      need the other products (eta,phi,mass of jet)
    #      then don't read them out. 
    ################################################

    ################################################
    #   Retrieve the jet vertex mass and plot the
    #   secondary vertex mass for tagged jets. 
    ################################################

    event.getByLabel( jetLabel, jetHandle )
    jets = jetHandle.product()

    # Find the njet bin we're in
    njets = 0
    # Now loop over the jets, and count tags
    ntags = 0
    jets_p4 = []
    for ijet in range( 0, len( jets ) ) :
        if jets[ijet].pt() > 30.0 :
            njets += 1
        ijetP4 = ROOT.TLorentzVector()
        ijetP4.SetPtEtaPhiM( jets[ijet].pt(), jets[ijet].eta(), jets[ijet].phi(), jets[ijet].mass() )
        jets_p4.append( ijetP4 )
        jetPtHist.Fill( jets[ijet].pt() )
        tagName = "pfCombinedInclusiveSecondaryVertexV2BJetTags"
        if jets[ijet].bDiscriminator(tagName) > 0.890:
            ntags = ntags + 1
            jetSecvtxMass = jets[ijet].vtxMass()
            secvtxMassHist.Fill( jetSecvtxMass )
            if options.doMC :
                if abs(jets[ijet].partonFlavour()) == 5 :
                    secvtxMassHistB.Fill( jetSecvtxMass )
                elif abs(jets[ijet].partonFlavour()) == 4 :
                    secvtxMassHistC.Fill( jetSecvtxMass )
                else :
                    secvtxMassHistL.Fill( jetSecvtxMass )
 
    # We're not interested in <=4 jets
    if njets < minJets :
        continue
    nEventsPassed4Jets = nEventsPassed4Jets + 1

    if ntags > 0:
        nEventsPassed1Tag = nEventsPassed1Tag + 1
        pairs.append( [event.object().id().run(),
                        event.object().id().luminosityBlock(),
                        event.object().id().event(),
                        njets,
                        ntags] )
    else :
        continue


    ################################################
    #   Require exactly one lepton (e or mu)
    #   ------------------------------------
    #      Our ntuples have both muon and electron
    #      events, and hence we must select events
    #      based on one or the other type. 
    #      To accomplish this we check the products
    #      for the type we're currently plotting
    #      (Mu or Ele), and check if the product is
    #      present. 
    ################################################
    event.getByLabel (muonLabel, muonHandle)
    muons = muonHandle.product()
    event.getByLabel (electronLabel, electronHandle)
    electrons = electronHandle.product() 

    numMuons = len(muons)
    numElectrons = len(electrons)

    # If neither muons nor electrons are found, skip
    if numMuons == 0 and numElectrons == 0 :
        continue
    # If we are looking for muons but none are found, skip
    if options.lepType == 0 and numMuons == 0 :
        continue
    # If we are looking for electrons but none are found, skip
    if options.lepType == 1 and numElectrons == 0 :
        continue

    # keep leptons with certain pt threshold in the event
    if options.lepType == 0 and muons[0].pt() <= muonPtMin:
        continue
    if options.lepType == 1 and electrons[0].pt() <= electronPtMin:
        continue
    
    # Now get the MET
    event.getByLabel( metLabel, metHandle )
    met = metHandle.product()[0].pt()

    # Now get the PF isolation
    lepIso = -1.0
    if options.lepType == 0 :
        nhIso = muons[0].neutralHadronIso()
        chIso = muons[0].chargedHadronIso()
        phIso = muons[0].photonIso() 
        puIso = muons[0].puChargedHadronIso()  
        lepIso = (chIso + max(0.0, nhIso + phIso - 0.5*puIso)) / muons[0].pt()
    if options.lepType == 1 :
        nhIso = electrons[0].neutralHadronIso()
        chIso = electrons[0].chargedHadronIso()
        phIso = electrons[0].photonIso()        
        puIso = electrons[0].puChargedHadronIso() 
        lepIso = (chIso + max(0.0, nhIso + phIso - 0.5*puIso)) / electrons[0].pt()

    # Make a plot of the MET versus ISO for normalization purposes
    metVsIso.Fill( met, lepIso )

    # If the MET is lower than our cut, skip, unless we want it inverted
    if not options.invertMET :
        if met < metMin :
            continue
    else :
        if met > metMin :
            continue

    # If the ISO is higher than our cut, skip, unless we want it inverted
    if not options.invertPFIso :
        if lepIso > isoMax :
            continue
    else :
        if lepIso < isoMax :
            continue

    # Now compute m3
    maxPt = -1.0
    m3 = -1.0
    for ijet in range(0, len(jets_p4) ) :
        for jjet in range(ijet + 1, len(jets_p4) ) :
            for kjet in range(jjet + 1, len(jets_p4) ) :
                sumP4 = jets_p4[ijet] + jets_p4[jjet] + jets_p4[kjet]
                if sumP4.Perp() > maxPt :
                    maxPt = sumP4.Perp()
                    m3 = sumP4.M()
    if maxPt > 0.0 :
        m3Hist.Fill( m3 )

# Stop our timer
timer.Stop()

# Print out our timing information
rtime = timer.RealTime(); # Real time (or "wall time")
ctime = timer.CpuTime(); # CPU time
print("Analyzed events: {0:6d}").format(nEventsAnalyzed)
print(">=4 jet events : {0:6d}").format(nEventsPassed4Jets)
print(">=1 tag events : {0:6d}").format(nEventsPassed1Tag)
print("RealTime={0:6.2f} seconds, CpuTime={1:6.2f} seconds").format(rtime,ctime)
print("{0:4.2f} events / RealTime second .").format( nEventsAnalyzed/rtime)
print("{0:4.2f} events / CpuTime second .").format( nEventsAnalyzed/ctime)



f.cd()
f.Write()
f.Close()

File: TopAnalysis/test/topAnalysis.cc
#include "DataFormats/FWLite/interface/Handle.h"
#include "DataFormats/FWLite/interface/Event.h"
#include "DataFormats/FWLite/interface/ChainEvent.h"
#include "CATTools/DataFormats/interface/Muon.h"
#include "CATTools/DataFormats/interface/Electron.h"
#include "CATTools/DataFormats/interface/Jet.h"
#include "CATTools/DataFormats/interface/MET.h"
#include "TROOT.h"
#include "TFile.h"
#include "TStopwatch.h"
#include "TH1F.h"
#include "TH2F.h"
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

using namespace fwlite;
using namespace std;
const double pi = TMath::Pi();

void topAnalysis(std::string fileNamesIn, std::string outputFile="top_fwlite.root",
                 bool doMC=false, int lepType=0,
                 bool invertMET=false, bool invertPFIso=false,
                 int maxEvents=100000000)
{
  gROOT->Macro("rootlogon.C");

  TFile* f = TFile::Open("out.root", "recreate");
  f->cd();

  // Book histograms
  auto secvtxMassHist = new TH1F("secvtxMassHist", "Secondary Vertex Mass", 150, 0., 5.0);
  auto secvtxMassHistB = new TH1F("secvtxMassHistB", "Secondary Vertex Mass, b", 150, 0., 5.0);
  auto secvtxMassHistC = new TH1F("secvtxMassHistC", "Secondary Vertex Mass, c", 150, 0., 5.0);
  auto secvtxMassHistL = new TH1F("secvtxMassHistL", "Secondary Vertex Mass, usdg", 150, 0., 5.0);
  auto metVsIso = new TH2F("metVsIso", "MET Versus PFIsolation", 15, 0., 150., 200, 0., 4.0);
  auto jetPtHist = new TH1F("jetPtHist", "Jet p_{T}", 150, 0., 600.);
  auto m3Hist = new TH1F("m3Hist", "M3 Histogram", 150, 0., 600.);
  auto nJetsHist = new TH1F("nJetsHist","Number of Jets", 10, 0, 10);

  // Kinematic cuts
  const double jetPtMin = 30., leadJetPtMin = 30.;
  const double isoMax = 0.2, ssvheCut = 1.74;
  const int minJets = 4;
  const double muonPtMin = (lepType == 0 ? 45 : 20);
  const double electronPtMin = (lepType == 0 ? 20 : 35);
  const double metMin = 20;

  std::vector<std::string> files;
  if ( fileNamesIn.rfind(".root") != std::string::npos )
  {
    files.push_back(fileNamesIn);
  }
  else
  {
    std::ifstream inputFiles(fileNamesIn);
    while ( true ) 
    {
      std::string fName;
      if ( ! (inputFiles >> fName) ) break;
      if ( fName.empty() ) continue;
      files.push_back(fName);
    }
  }
  fwlite::ChainEvent events(files);

  int nEventsAnalyzed = 0, nEventsPassed4Jets = 0, nEventsPassed1Tag = 0;
  TStopwatch timer;
  timer.Start();

  int nTotal = events.size();
  if ( nTotal > maxEvents ) {
    cout << "Sample is too large(" << nTotal << ")! number of Total event is changed to " << maxEvents << endl;
    nTotal = maxEvents;
  }
  float ipercentDoneLast = -1, sumTime = 0;
  cout << "Start looping\n";
  for ( events.toBegin(); !events.atEnd(); ++events )
  {
    if ( nEventsAnalyzed == nTotal ) break;

    ++nEventsAnalyzed;
    const double percentDone = float(nEventsAnalyzed)/float(nTotal)*100.0;
    const int ipercentDone(percentDone);
    if ( ipercentDone != ipercentDoneLast ) {
      const double sumTime = timer.RealTime();
      timer.Start();
      ipercentDoneLast = ipercentDone;
      cout << "Processing " << nEventsAnalyzed << "/" << nTotal << " : " << ipercentDone << "%\n";
    }
    
    const edm::EventBase& event(events);

    edm::Handle<std::vector<double> > puHandle;
    edm::Handle<std::vector<cat::Jet> > jetHandle;
    edm::Handle<std::vector<cat::Muon> > muonHandle;
    edm::Handle<std::vector<cat::Electron> > electronHandle;
    edm::Handle<std::vector<cat::MET> > metHandle;

    event.getByLabel(edm::InputTag("pileupWeight"), puHandle);
    event.getByLabel(edm::InputTag("catJets::CAT"), jetHandle);
    event.getByLabel(edm::InputTag("catMuons::CAT"), muonHandle);
    event.getByLabel(edm::InputTag("catElectrons::CAT"), electronHandle);
    event.getByLabel(edm::InputTag("catMETs::CAT"), metHandle);

    const double met = metHandle->at(0).pt();

    TLorentzVector lep_p4;
    int nMuon = 0, nElectron = 0;
    double lep_pt = 0, lep_eta = 0, lep_phi = 0, lep_iso = 0;
    for ( unsigned int imu = 0; imu < muonHandle->size(); ++imu ) {
      const auto& mu = muonHandle->at(imu);
      const double pt = mu.pt();
      const double eta = mu.eta();
      if ( mu.isTightMuon() and mu.pt() > muonPtMin and std::abs(eta) < 2.5 ) {
        ++nMuon;
        if ( lepType == 1 and mu.pt() > lep_pt ) {
          lep_pt = pt;
          lep_eta = eta;
          lep_phi = mu.phi();
          const double nhIso = mu.neutralHadronIso();
          const double chIso = mu.chargedHadronIso();
          const double phIso = mu.photonIso();
          const double puIso = mu.puChargedHadronIso();
          lep_iso = (chIso + max(0.0, nhIso + phIso - 0.5*puIso)) / pt;
          lep_p4.SetPxPyPzE(mu.px(), mu.py(), mu.pz(), mu.energy());
        }
      }
    }
    for  ( unsigned int iel = 0; iel < electronHandle->size(); ++iel ) {
      const auto& el = electronHandle->at(iel);
      const double pt = el.pt();
      const double eta = el.eta();
      if ( el.isPF() and el.passConversionVeto() and pt > electronPtMin and std::abs(eta) < 2.5 ) {
        ++nElectron;
        if ( lepType == 0 and pt > lep_pt ) {
          lep_pt = pt;
          lep_eta = eta;
          lep_phi = el.phi();
          const double nhIso = el.neutralHadronIso();
          const double chIso = el.chargedHadronIso();
          const double phIso = el.photonIso();
          const double puIso = el.puChargedHadronIso();
          lep_iso = (chIso + max(0.0, nhIso + phIso - 0.5*puIso)) / pt;
        }
      }
    }
    const int nLepton = nMuon + nElectron;
    if ( nLepton != 1 ) continue;
    if ( lepType == 0 and nMuon == 0 ) continue;
    if ( lepType == 1 and nElectron == 0 ) continue;

    metVsIso->Fill(met, lep_iso);

    int nbjet = 0;
    std::vector<double> jets_pt, jets_eta, jets_phi, jets_btag, jets_svMass;
    std::vector<int> jets_flav;
    double evt_maxbtag = -1, evt_svMassAtMaxbtag = -1;
    for ( unsigned int ijet = 0; ijet < jetHandle->size(); ++ijet ) {
      const auto& jet = jetHandle->at(ijet);
      const double pt = jet.pt();
      const double eta = jet.eta();
      if ( pt < 30 or std::abs(eta) > 2.5 ) continue;
      if ( std::hypot(eta-lep_eta, jet.phi()-lep_phi) < 0.4 ) continue;
      
      const double btag = jet.bDiscriminator("pfCombinedInclusiveSecondaryVertexV2BJetTags");
      const int flav = (doMC ? 0 : std::abs(jet.partonFlavour()));

      jets_pt.push_back(pt);
      jets_eta.push_back(eta);
      jets_phi.push_back(jet.phi());
      jets_btag.push_back(btag);
      jets_svMass.push_back(-1);
      jets_flav.push_back(flav);

      if ( btag > 0.890 ) {
        ++nbjet;
        jets_svMass.back() = jet.vtxMass();
        if ( doMC ) {
          switch ( std::abs(flav) ) {
            case 5: secvtxMassHistB->Fill(jets_svMass.back()); break;
            case 4: secvtxMassHistC->Fill(jets_svMass.back()); break;
            default: secvtxMassHistL->Fill(jets_svMass.back()); break;
          }
        }
      }
      if ( evt_maxbtag < btag ) {
        evt_maxbtag = btag;
        evt_svMassAtMaxbtag = jets_svMass.back();
      }

      jetPtHist->Fill(pt);
      secvtxMassHist->Fill(jets_svMass.back());
    }
    const int njets = jets_pt.size();
    nJetsHist->Fill(njets);
    if ( njets < minJets ) continue;
    ++nEventsPassed4Jets;

    if ( nbjet == 0 ) continue;
    ++nEventsPassed1Tag;

    if ( invertMET and met > metMin ) continue;
    else if ( !invertMET and met < metMin ) continue;

    if ( invertPFIso and lep_iso < isoMax ) continue;
    else if ( !invertPFIso and lep_iso > isoMax ) continue;

    // Now compute m3
    const double m3 = (jetHandle->at(0).p4()+jetHandle->at(1).p4()+jetHandle->at(2).p4()).mass();
    m3Hist->Fill(m3);
  }

  f->cd();
  secvtxMassHist->Write();
  secvtxMassHistB->Write();
  secvtxMassHistC->Write();
  secvtxMassHistL->Write();

  metVsIso->Write();
  jetPtHist->Write();
  m3Hist->Write();
  nJetsHist->Write();

  f->Write();

}
